tree-sitter/src/TreeSitter/Unmarshal.hs

```diff
@@ -46,52 +47,55 @@ parseByteString language bytestring = withParser language $ \ parser -> withPars
+  unmarshalNodes :: (MonadFail m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => [Node] -> m a
+  default unmarshalNodes :: (MonadFail m, Carrier sig m, GUnmarshal (Rep a), Generic a, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => [Node] -> m a
+  unmarshalNodes [x] = do
+    goto (nodeTSNode x)
+    to <$> gunmarshalNode
+  unmarshalNodes [] = fail "expected a node but didn't get one"
+  unmarshalNodes _ = fail "expected a node but got multiple"
 
 instance Unmarshal Text.Text where
-  unmarshalNode = do
-    node <- peekNode
-    case node of
-      Just node' -> do
-        bytestring <- ask
-        let start = fromIntegral (nodeStartByte node')
-            end = fromIntegral (nodeEndByte node')
-        pure (decodeUtf8 (slice start end bytestring))
-      _ -> fail "expected a node containing Text but didn't get one"
+  unmarshalNodes [node] = do
+    bytestring <- ask
+    let start = fromIntegral (nodeStartByte node)
+        end = fromIntegral (nodeEndByte node)
+    pure (decodeUtf8 (slice start end bytestring))
+  unmarshalNodes [] = fail "expected a node but didn't get one"
+  unmarshalNodes _ = fail "expected a node but got multiple"
 
 instance Unmarshal a => Unmarshal (Maybe a) where
-  unmarshalNode = Just <$> unmarshalNode
-  unmarshalEmpty = pure Nothing
+  unmarshalNodes [] = pure Nothing
+  unmarshalNodes listOfNodes = Just <$> unmarshalNodes listOfNodes
 
 instance (Unmarshal a, Unmarshal b, SymbolMatching a, SymbolMatching b) => Unmarshal (Either a b) where
-  unmarshalNode = do
-    currentNode <- peekNode
-    (lhsSymbolMatch, rhsSymbolMatch, currentNode) <- case currentNode of
-      Just node -> pure (symbolMatch (Proxy @a) node, symbolMatch (Proxy @b) node, node)
-      Nothing -> fail "expected a node of type (Either a b) but didn't get one"
+  unmarshalNodes [node] = do
+    let lhsSymbolMatch = symbolMatch (Proxy @a) node
+        rhsSymbolMatch = symbolMatch (Proxy @b) node
     if lhsSymbolMatch
-      then Left <$> unmarshalNode @a
+      then Left <$> unmarshalNodes @a [node]
       else if rhsSymbolMatch
-        then Right <$> unmarshalNode @b
-        else fail $ showFailure (Proxy @(Either a b)) currentNode
+        then Right <$> unmarshalNodes @b [node]
+        else fail $ showFailure (Proxy @(Either a b)) node
+  unmarshalNodes [] = fail "expected a node of type (Either a b) but didn't get one"
+  unmarshalNodes _ = fail "expected a node of type (Either a b) but got multiple"
+
 
 instance Unmarshal a => Unmarshal [a] where
-  -- FIXME: This is wrong. Repeated fields are represented in the tree as multiple nodes with the same field name.
-  --        Currently we only represent a single node for each field name,
-  --        so we only end up keeping the last one encountered in the tree.
-  unmarshalNode = pure <$> unmarshalNode
-  unmarshalEmpty = pure []
+  unmarshalNodes (x:xs) = do
+    goto (nodeTSNode x)
```

Comment:
Is this `goto` still necessary?

Reply:
No. The call to `unmarshalNodes` takes care of it.